lib/dialects/abstract/query-generator.js

```diff
@@ -1194,12 +1194,12 @@ var QueryGenerator = {
             , tableRight = as
             , attrRight = association.associationType !== 'BelongsTo' ?
                           association.identifierField || association.identifier :
-                          right.rawAttributes[primaryKeysRight[0]].field || primaryKeysRight[0]
+                          right.rawAttributes[association.targetIdentifier || primaryKeysRight[0]].field
```

Comment:
I couldn't completely boil this down to `association.targetIdentifier` since this code is called from a `belongsToMany` relationship as well, which apparently doesn't leverage `association.targetIdentifier`.